controllers/datasource_controller.go

```diff
@@ -72,6 +66,27 @@ func (r *DatasourceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		}
 		return reconcile.Result{}, err
 	}
+
+	if instance.DeletionTimestamp != nil {
+		logger.Info("Delete datasource")
+
+		// remove the finalizer to complete the delete action
+		instance.Finalizers = utils.RemoveString(instance.Finalizers, arcadiav1alpha1.Finalizer)
+		err := r.Client.Update(ctx, instance)
+		if err != nil {
+			logger.Error(err, "Failed to remove repo finalizer")
+		}
```

Comment:
**Suggestion:** It seems like you are using a logger to log errors before returning them. This can lead to duplicate log entries if the caller also logs the error. It is a good practice to either log the error or return it, but not both. You should remove the logger calls before the error returns.
```suggestion
  if err != nil {
    return reconcile.Result{}, err
  }
```